drv/lpc55-swd/src/armv7debug.rs

```diff
@@ -0,0 +1,189 @@
+        // At least one reset since last DHCSR read. clear on read.
+        const S_RESET_ST = 1 << 25;
+        const S_RETIRE_ST = 1 << 24;
+        const S_LOCKUP = 1 << 19;
+        const S_SLEEP = 1 << 18;
+        const S_HALT = 1 << 17;
+        const S_REGRDY = 1 << 16;
+
+        // Magic number allows setting C_* bits.
+        const DBGKEY = 0xA05F << 16;
+
+        const C_SNAPSTALL = 1 << 5;
+        const C_MASKINTS = 1 << 3;
+        const C_STEP = 1 << 2;
+        const C_HALT = 1 << 1;
+        const C_DEBUGEN = 1 << 0;
+        const _ = !0;
+    }
+}
+
+impl From<u32> for Dhcsr {
+    fn from(v: u32) -> Self {
+        Self::from_bits_retain(v)
+    }
+}
+
+impl DpAddressable for Dhcsr {
+    fn addr() -> u32 {
+        Self::ADDRESS
+    }
+}
+
+impl Dhcsr {
+    const ADDRESS: u32 = 0xE000EDF0;
+    pub fn halt() -> Self {
+        Self::DBGKEY | Self::C_HALT | Self::C_DEBUGEN
+    }
+    pub fn resume() -> Self {
+        Self::DBGKEY | Self::C_DEBUGEN
+    }
+    pub fn end_debug() -> Self {
+        Self::DBGKEY
+    }
+    pub fn is_halted(self) -> bool {
+        self & Self::S_HALT == Self::S_HALT
+    }
+    pub fn is_regrdy(self) -> bool {
+        self & Self::S_REGRDY == Self::S_REGRDY
+    }
+    pub fn _is_lockup(self) -> bool {
+        self & Self::S_LOCKUP == Self::S_LOCKUP
+    }
+}
+
+// Debug Core Register Selector Register
+pub const DCRSR: u32 = 0xE000EDF4;
+// Debug Core Register Data Register
+pub const DCRDR: u32 = 0xE000EDF8;
+
+// DEMCR RW   0x00000000    Debug Exception and Monitor Control Register
+bitflags! {
+    #[derive(PartialEq, Eq, Copy, Clone)]
+    pub struct Demcr: u32 {
+        const MON_EN = 1 << 16;
+        const VC_HARDERR = 1 << 10;
+        const VC_INTERR = 1 << 9;
+        const VC_BUSERR = 1 << 8;
+        const VC_STATERR = 1 << 7;
+        const VC_CHKERR = 1 << 6;
+        const VC_NOCPERR = 1 << 5;
+        const VC_MMERR = 1 << 4;
+        const VC_CORERESET = 1 << 0;
+    }
+}
+
+impl DpAddressable for Demcr {
+    fn addr() -> u32 {
+        Self::ADDRESS
+    }
+}
+
+impl Demcr {
+    const ADDRESS: u32 = 0xE000EDFC;
+}
+
+// Armv7-M Arch. Ref. Manual - C1.6.1 Debug Fault Status Register
+// RW   Init: 0x00000000[on power-on reset only]
+bitflags! {
+    #[derive(PartialEq, Eq, Copy, Clone)]
+    pub struct Dfsr: u32 {
+        // Assertion of an external debug request
+        const EXTERNAL = 1 << 4;
+        // Vector catch triggered
+        const VCATCH = 1 << 3;
+        // At least one DWT event
+        const DWTTRAP = 1 << 2;
+        // Breakpoint
+        const BKPT = 1 << 1;
+        // Halt request debug event.
+        // • A C_HALT or C_STEP request, triggered by a write to the DHCSR,
+        //   see Debug Halting Control and Status Register, DHCSR.
+        // • A step request triggered by setting DEMCR.MON_STEP to 1,
+        //   see Debug monitor stepping on page C1-696.
+        const HALTED = 1 << 0;
+        const _ = !0;
```

Comment:
Is this a quirk of bitflags?

Reply:
Yes. It avoids surprises with unspecified bits.

See https://docs.rs/bitflags/latest/bitflags/#externally-defined-flags